feat: Adds support to multiple dependencies to the native filters

michael-s-molina opened the pull request:
### SUMMARY
Adds support to multiple dependencies to the native filters. This will allow the user to create hierarchical and graph relationships among filters. The changes are the result of a series of interviews conducted by the design team with Superset users. Here's a summary of the changes:
- Creates the "Values are dependent on other filters" section
- Removes the "Filters is hierarchical" section
- Adds indentation to the collapsable containers
- Removes the hierarchical display of filters in the left panel of the modal
- Adjusts the algorithm to validate circular dependencies on the fly
- [Remove cascading popovers from filter bar](https://github.com/apache/superset/pull/18790) @kgabryje
- Effectively uses the [filter cards](https://github.com/apache/superset/pull/18874) to display filter context
- Adjusts RTL and Cypress the tests

It's important to note that major refactors are not in the scope at this stage. Our first objective is to deprecate FilterBox. We'll tackle code improvements in the next stage of the native filters revamp.

@kasiazjc @jess-dillard

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/155568831-ae7871b9-5c54-4060-beeb-11ac70978463.mov


### TESTING INSTRUCTIONS
We need to test the whole module. Apart from the basic tests, some important behaviors that should be checked:
- When selecting "Values are dependent on other filters" a value is automatically suggested and this value does not generate a circular dependency warning if possible.
- The circular dependency checking should occur on the fly while the user is selecting the filter dependencies.
- If a filter is removed or has its type changed to one different than "select", all references to this filter should display "(deleted or invalid filter)". After saving and reopening the modal, that reference should disappear.
- Only filters of type "select" should be available to be referenced as a dependency.
- All filters in scope should be displayed on the filter bar independently of their relationships.
- The values of a filter should be affected by its dependencies when applicable.
- The filter card context is correctly displayed and text is truncated as expected. Please refer to #18874 to see the requirements.

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

kgabryje committed:
chore(native-filters): Remove cascading popovers from filter bar

kgabryje commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigModal.test.tsx:
```diff
@@ -335,7 +327,7 @@ test.skip("doesn't render time range pre-filter if there are no temporal columns
 });
 
-test('filter title groups are draggable', async () => {
+test('filtes are draggable', async () => {
```
```suggestion
test('filters are draggable', async () => {
```
😛

michael-s-molina replied:
🙌🏼

kgabryje committed:
Fix type errors, change parentFilters to dependencies

michael-s-molina commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DependencyList.tsx:
```diff
@@ -0,0 +1,198 @@
+  };
+
+  const onChange = (id: string, value: string) => {
+    const indexOf = rows.findIndex(e => e === id);
```
I think this check is not necessary because it will always find the id.

kgabryje commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DependencyList.tsx:
```diff
@@ -0,0 +1,202 @@
+
+  const onAdd = () => {
+    const filter = availableFilters.find(
+      availableFilter => !rows.includes(availableFilter.value),
```
Improvement suggestion - don't autoselect a filter that leads to circular dependency (but leave it in the list of availableFilters). For example, if we have filters A, B and C, and A depends on B, when user tries to add dependency in filter B, autoselect C instead of A

michael-s-molina replied:
That would be nice indeed. I'll work on this.

michael-s-molina replied:
I implemented your suggestion.

kgabryje commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DependencyList.tsx:
```diff
@@ -0,0 +1,202 @@
+          'Values selected in other filters will affect the filter options to only show relevant values',
+        )}
+      >
+        {hasDependencies && <Label>{t('Values dependent on')}</Label>}
```
This text has color #666666, while in the designs it's #b2b2b2

kgabryje commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/utils.ts:
```diff
@@ -25,40 +25,48 @@ import {
+  if (dependencies) {
+    let result = false;
+    dependencies.forEach(dependency => {
+      result ||= hasCircularDependency(form, filterConfigMap, dependency, [
```
Maybe we should return once result is true instead of checking all dependencies?
Like:
```
if (dependencies) {
    return dependencies.some(dependency => hasCircularDependency(form, filterConfigMap, dependency, [
        ...trace,
        filterId,
      ]));
}

kgabryje commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/utils.ts:
```diff
@@ -241,61 +172,23 @@ export const createHandleRemoveItem =
-      setFilterHierarchy(prevFilterHierarchy =>
-        buildNewFilterHierarchy(prevFilterHierarchy),
+      setOrderedFilters((orderedFilters: string[]) =>
+        orderedFilters.filter(filter => !isEqual(filter, filterId)),
```
```suggestion
        orderedFilters.filter(filter => filter !== filterId),
```
`isEqual` feels like a bit of an overkill for comparing 2 strings 🙂

michael-s-molina replied:
Oh man... my Java background gets me every time on this 😝. You should see me when I'm writing an `if` and thinking about Javascript, Java, and Python at the same time! It's hilarious 😆

kgabryje committed:
Change variable name

kgabryje commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigModal.tsx:
```diff
@@ -369,11 +386,34 @@ export function FiltersConfigModal({
     setOrderedFilters(newOrderedFilter);
   };
 
+  const buildDependencyMap = useMemo(
```
If you use `useCallback`, you can get rid of one of `() =>`
Like
```
const buildDependencyMap = useCallback(() => {
   ...
})
```

michael-s-molina replied:
But I also want to cache the result and not just the function.

michael-s-molina committed:
feat: Adds support to multiple dependencies to the native filters

kgabryje committed:
Attach filter cards to native filters in filter bar

michael-s-molina committed:
Enforces type

kgabryje committed:
Fix typing

kgabryje committed:
A few enhancements to filter card

kgabryje committed:
Add scope tooltip sections

kgabryje committed:
Fix filtering by dependency

michael-s-molina commented:
/testenv up

michael-s-molina commented:
@kasiazjc @jinghua-qa @rusackas We can start testing the feature. Don’t worry about code review, this PR is a combination of a few PRs that have already been reviewed.

As soon as we get Preset’s approval, we’ll tag more people from the community to also validate the feature.

michael-s-molina commented:
/testenv up

kgabryje commented:
I found a problem in config modal - when we add a new filter, dependencies for other filters get reset

https://user-images.githubusercontent.com/15073128/155576669-6aece1ab-2b76-48b4-81b2-b2a9cb43ec7a.mov

michael-s-molina committed:
Fixes dependency reset

michael-s-molina commented:
> @michael-s-molina I found a problem in config modal - when we add a new filter, dependencies for other filters get reset

Fixed.

> Also, it looks like adding a new filter triggers rerendering of all other filters and charts, but I'm not sure if this was introduced in this PR or if it's an old issue

This also happens on master so it's not a problem introduced by this PR. I think this falls under the category of performance improvements and should be planned and addressed in another PR.

michael-s-molina commented:
/testenv up

jinghua-qa commented:
I found an issue that default value did not stay

https://user-images.githubusercontent.com/81597121/155784379-c0b263b2-695e-4483-9992-24dfbf7be02b.mov

jinghua-qa commented:
Another issue that could be related to the default value did not show in the filter is:
when select time range with required setting, since the default value did not show in time range filter, it show error in time range filter and other filters can not apply change

https://user-images.githubusercontent.com/81597121/155816507-4a9be646-7d16-4aad-8b54-519cd86a7969.mov

jinghua-qa commented:
When user select a scope for the filter, the scope of the chart information will not show in the tooltip? When the filter apply to all it will show " apply to all charts", but if it is only applied to selected chart or tab it did not show which scope it is in the tooltip, is this expected?


Uploading Screen Recording 2022-02-25 at 3.39.51 PM.mov…

kgabryje commented:
> When user select a scope for the filter, the scope of the chart information will not show in the tooltip? When the filter apply to all it will show " apply to all charts", but if it is only applied to selected chart or tab it did not show which scope it is in the tooltip, is this expected?
> 
>  Screen.Recording.2022-02-25.at.3.33.54.PM.mov

Yes - tooltip shows only when text is truncated. I think there's no point in listing every chart in the dashboard when scope is "All charts". As for the second case - yes, not showing "Scope" row when nothing is in scope was intentional. I can add a row "Scope: None", but I don't think there's a point in creating a filter with nothing in scope and then opening "Out of scope panel" to see that filter. @kasiazjc wdyt?

kasiazjc commented:
> > When user select a scope for the filter, the scope of the chart information will not show in the tooltip? When the filter apply to all it will show " apply to all charts", but if it is only applied to selected chart or tab it did not show which scope it is in the tooltip, is this expected?
> > Screen.Recording.2022-02-25.at.3.33.54.PM.mov
> 
> Yes - tooltip shows only when text is truncated. I think there's no point in listing every chart in the dashboard when scope is "All charts". As for the second case - yes, not showing "Scope" row when nothing is in scope was intentional. I can add a row "Scope: None", but I don't think there's a point in creating a filter with nothing in scope and then opening "Out of scope panel" to see that filter. @kasiazjc wdyt?

For "All charts" we won't be showing a tooltip, because as Kamil said it does not make sense to list all of the charts/tabs. If there is no scope added though we should still keep the "scope" row with value "none" as it is a valuable information for the user who is looking for information about the filter.

kgabryje commented:
> I found an issue that default value did not show in the filter value, the filter is applied, i saw it in the query and in the filter indicator.
> 
>  Screen.Recording.2022-02-25.at.11.39.42.AM.mov

Thanks for spotting that! It looks like the bug is present on master though, so maybe let's fix that in a separate PR?
Same applies to the second comment about filter value

kgabryje commented:
@jinghua-qa Fixed by #18963

kgabryje committed:
Show empty scope as "None"

kgabryje commented:
Empty scope done @jinghua-qa @kasiazjc 

<img width="521" alt="image" src="https://user-images.githubusercontent.com/15073128/155995505-b6b87f01-47d3-4a93-9b44-2927645dd397.png">

jinghua-qa commented:
> > > When user select a scope for the filter, the scope of the chart information will not show in the tooltip? When the filter apply to all it will show " apply to all charts", but if it is only applied to selected chart or tab it did not show which scope it is in the tooltip, is this expected?
> > > Screen.Recording.2022-02-25.at.3.33.54.PM.mov
> > 
> > 
> > Yes - tooltip shows only when text is truncated. I think there's no point in listing every chart in the dashboard when scope is "All charts". As for the second case - yes, not showing "Scope" row when nothing is in scope was intentional. I can add a row "Scope: None", but I don't think there's a point in creating a filter with nothing in scope and then opening "Out of scope panel" to see that filter. @kasiazjc wdyt?
> 
> For "All charts" we won't be showing a tooltip, because as Kamil said it does not make sense to list all of the charts/tabs. If there is no scope added though we should still keep the "scope" row with value "none" as it is a valuable information for the user who is looking for information about the filter.

Thank you for your explanation.

michael-s-molina commented:
Hi @graceguo-supercat @amitmiran137. I'm tagging you to help us review this PR that effectively implements all the changes planned for stage 1 of native filters. This PR is a combination of multiple PR that should be merged together and were previously reviewed so you can focus more on functionality than the code. Tagging @rusackas as well who is coordinating the work for deprecating FilterBox which should be possible after this PR.

michael-s-molina commented:
/testenv up

jinghua-qa commented:
I found an issue related to the number filter, it is also in master
https://github.com/apache/superset/issues/18985

kgabryje committed:
Remove old test

jinghua-qa reviewed (approved):
Tested PRs with the new changes and also executed some regression test, did not find any blocker. Here is the test plan and result: https://docs.google.com/spreadsheets/d/1KbhE_FQzw38rHwyNGbQnJY0ljCZOVadgwpCQWAsNq7M/edit#gid=0

Have found some native filter related issues that existing in master, would love to see them fixed in other PRs:
https://github.com/apache/superset/issues/19000
https://github.com/apache/superset/issues/18985

kgabryje committed:
Fix filter card not disappearing in some cases

kgabryje committed:
Dont underline the separators in dependency row

kgabryje commented:
/testenv up

kasiazjc commented:
Looks great! From my perspective - all of the things that were supposed to be included in Stage 1 were implemented and are added to this PR. I also tested the feature and it seems to be working well. Kudos @michael-s-molina @kgabryje @jinghua-qa 👏 

One additional comment from me - you may see that the filter type is different on filter card and in "add/edit filters" modal. This is something that has to be fixed separately and hopefully will be fixed soon (cc: @kgabryje)

kgabryje commented:
> One additional comment from me - you may see that the filter type is different on filter card and in "add/edit filters" modal. This is something that has to be fixed separately and hopefully will be fixed soon (cc: @kgabryje)

The reason for that is that in filter cards we use "official" filter names from metadata, while in the modal there's a mapping from metadata names to different names. We should change the names in metadata to match the ones in the mapping, probably do a db migration and delete a mapping.
CC @villebro

graceguo-supercat commented:
@michael-s-molina this is great! i play with it sample dashboards, no issues were found!
I have only 1 question. definitely LGTM!

graceguo-supercat reviewed (approved):
LGTM

graceguo-supercat commented on superset-frontend/src/dashboard/components/DashboardBuilder/DashboardContainer.tsx:
```diff
@@ -91,9 +91,9 @@ const DashboardContainer: FC<DashboardContainerProps> = ({ topLevelTabs }) => {
       const chartsInScope: number[] = getChartIdsInFilterScope({
         filterScope: {
-          scope: scope.rootPath,
+          scope: scope?.rootPath,
```
In what cases `scope` field could be optional? I thought all filter has default "all" scope and [] for immune.

kgabryje committed:
Remove ?. operator from scope
